### PR TITLE
[Feature][Sample] Add Triton apply_top_k_top_p fused op for sampling

### PR DESCRIPTION
**Pull request by 15625122759**
## What

A Triton implementation of top-k/top-p filtering for the sampler, opt-in via `VLLM_ASCEND_USE_TRITON_APPLY_TOPK_TOPP=1` (default off). Two entry points:

- `apply_top_k_top_p` — masked logits, bit-exact with `torch_npu.npu_top_k_top_p` (`torch.equal` on fp32/bf16/fp16)
- `fused_topk_topp_softmax` — same filtering with the final softmax folded into the kernel, returns fp32 probs directly; this is the branch the sampler uses when logprobs are not requested

Both take the per-request `[B]` k/p tensors from `_get_topk_topp_constraints()` as-is. The CANN op forces a `.item()` on that tensor (host sync on every decode step); this op doesn't.

## How it works

The host does the sort once (`torch.sort`, ascending, stable) and the kernel only works on the sorted row. Because the input is ascending, both the top-k mask (`s >= topk_value`) and the top-p mask (`cumsum > 1 - p`) are suffix-monotone, so the kept set collapses to a single scalar boundary `m = max(m_p, m_k)` per row. The kernel then only scatters the surviving suffix (~k elements) into a pre-filled output — the non-surviving positions are one contiguous block write instead of V random stores.

I used a scalar boundary rather than a cumsum-derived boolean vector as the scatter mask on purpose: triton-ascend 3.2.2 miscompiles that pattern (the store collapses to a few elements). The `v >= m` comparison mask sidesteps it.

kernel flow

Two kernel paths: V ≤ 4096 loads the row in one block; larger vocabularies go tiled — pass 1 accumulates the softmax denominator, pass 2 walks tiles until the cumulative probability crosses `1 - p` (early exit; `tl.cumsum` runs only inside the crossing tile), pass 3 only touches the kept suffix. `m_p`/`m_k` can be looser than the exact CANN cutoffs; the final `v >= m` mask restores exactness, so the output stays bit-exact.

The sampler wiring prefers logprobs-preserving paths: when `raw_logprobs` is requested it falls back to the masked-logits variant + separate softmax/log_softmax.

## Enabling it

With `VLLM_ASCEND_USE_TRITON_APPLY_TOPK_TOPP=1` the sampler logs once at startup:

switch log

## Performance

Test environment: single Ascend 910B3 (linux aarch64), CANN 9.0.0, torch 2.10.0 / torch_npu 2.10.0.post4, triton-ascend 3.2.2. Model: `Qwen3.8-27B-w8a8` (w8a8-quantized, vocab 151936). Serving benchmark command (identical for both runs):

```bash
vllm bench serve --model $MODEL_PATH --dataset-name random --random-input-len 1024 \
    --random-output-len 128 --num-prompts 300 --request-rate 10 \
    --top-p 0.95 --top-k 20 --temperature 0.6 --percentile-metrics ttft,tpot,itl
```

E2E serving benchmark, stock vs this PR (300/300 successful requests, identical total input/output tokens; 存量=stock, 改造后=this PR):

serving bench

Most of the win is removing the `npu_top_k_top_p` kernel from the graph — msprof side-by-side on the same model (Triton run vs stock baseline) shows its instruction count dropping ~20-24% on Cube/Vector/Computing units:

profiling

## Accuracy

triton acc
stock acc

## Tests

- `op/test_smoke.py` + `op/op_perf_test.sh`: 10-case matrix, fp32/bf16, single-pass + tiled paths, per-row k/p tensor and scalar inputs, k<0/k=0/k>V edge cases, determinism; bit-exact (`torch.equal`) vs `torch_npu.npu_top_k_top_p` + 20 randomized stress cases (N=1…1e5)
- `tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_apply_top_k_top_p.py`: 6 smoke cases incl. V=32000 tiled and per-row tensors
- ruff/mypy clean

Note on CI: `pre-commit` currently fails on `vllm_ascend/xlite/xlite.py:663` (`XModelConfig has no attribute index_full_mask`) — that's pre-existing on main since #16669; `tools/mypy.sh` type-checks the whole package so it fires on any PR. Nothing in this PR's files is flagged.

## Known limitations

- `tl.cumsum` lowers to an element-serial runtime call on triton-ascend 3.2.2 (~29.7us per 2048-wide tile vs ~0.2us for `tl.sum`, measured). The tiled kernel only runs it inside the single threshold-crossing tile, so the impact is small, but it would be nice to have a tree-scan lowering in the backend.
- The host-side `torch.sort` (~100-470us depending on B/V) stays. Callers that already hold sorted logits can use the `*_with_sorted` variants and skip it.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

File: vllm_ascend/ops/triton/apply_top_k_top_p.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+                idx = tl.load(si_ptr + row_base + offs, mask=kept, other=0)
+                tl.store(out_ptr + row_base + idx, tl.exp(s - s_max) * e_inv, mask=kept)
+        else:
+            # 散写原始 logit（masked logits，CANN 语义）
+            for t in range(t_start, V, BLOCK):
+                offs = t + offs_base
+                kept = (offs >= m) & (offs < V)
+                s = tl.load(sv_ptr + row_base + offs, mask=kept, other=0.0).to(tl.float32)
+                idx = tl.load(si_ptr + row_base + offs, mask=kept, other=0)
+                tl.store(out_ptr + row_base + idx, s, mask=kept)
```
[the remark this answers is not in the document]
Comment (15625122759) [reply vllm_ascend/ops/triton/apply_top_k_top_p.py]: Thanks for the careful review! These concerns are valid for generic Triton, but they don't apply here — this kernel already compiles and runs correctly on the target backend (triton-ascend 3.2.2, Ascend 910B3), and the suggested rewrite would be a significant performance regression:

1. **Compilation**: The tiled kernel (including the dynamic `if m_p == V:` branch in Pass 2, the `//` on a 0-d scalar, and the dynamic `t_start` loop bound) compiles and executes correctly on triton-ascend — verified bit-exact against `torch_npu.npu_top_k_top_p` (`torch.equal`) across fp32/bf16 and V=4097/32000/151936 (i.e., the tiled path), plus 20 randomized stress cases. Both constructs lower to uniform scalar control flow (`scf.if` / dynamic `scf.for` bounds), which the backend handles fine.

2. **Why Pass 2 keeps the early-exit**: `tl.cumsum` is lowered to an element-serial runtime call on triton-ascend 3.2.2 (~29.7us per BLOCK=2048 tile, vs ~0.2us for `tl.sum` — measured). The current guard ensures `tl.cumsum` executes **only in the single tile that crosses the p threshold**; the branchless rewrite would run it in **every** tile (74 tiles at V=151936), and would also keep scanning after the hit. For the typical decode case (k=20, p=0.95, threshold hit in tile 0) that's ~74x more cumsum work.

3. **Why Pass 3 keeps `t_start = (m // BLOCK) * BLOCK`**: with top-k first, the kept suffix typically spans only the last 1-2 tiles, so starting the scan at the floor-aligned `t_start` avoids re-reading the full vocabulary (e.g., 1 of 74 tiles at k=20, V=151936). Starting from 0 would turn Pass 3 back into a full-V scan (x2 loops in the fused variant).

The remaining risk the comment rightly points at (backend-specific scalar-CF handling) is covered by the added nightly NPU tests, which exercise this kernel end-to-end on the real backend.

**Comment by 15625122759** [comment]
**pre-commit status update** (after commit 0beb768):

All mypy errors in files touched by this PR are fixed:
- ~~`apply_top_k_top_p.py:97/158: "int" has no attribute "to"`~~ → `type: ignore[attr-defined]` (mypy infers the loop var as `int`; triton JIT handles `.to()` on the scalar)
- ~~`apply_top_k_top_p.py:400/437: Too many arguments`~~ → explicit unpack of `_sort_ascending()`

The remaining failure is **pre-existing upstream breakage unrelated to this PR**:

```
vllm_ascend/xlite/xlite.py:663: error: "XModelConfig" has no attribute "index_full_mask"  [attr-defined]
```

`index_full_mask` is set dynamically at runtime (`xlite.py:640`) and is not declared on `XModelConfig` (`xlite/utils.py:155`). Introduced by #16669 (merged 2026-09-20); since `tools/mypy.sh` runs mypy over the whole `vllm_ascend` package, this currently blocks pre-commit on **every** open PR. I'll open a separate one-line fix PR (getattr-based, consistent with the defensive access already used at `xlite.py:647`).

**Comment by 15625122759** [comment]
All code checks are green now (pre-commit incl. whole-package mypy for py3.10/3.11/3.12, cpu-ut 5711 cases, DCO).

`ci-gate` reports that precision/full tests are required for this PR since it touches source code. Could a maintainer please add the `ready-precise` label so the selected NPU tests run? Happy to address anything that comes out of them.

For reviewers' convenience, single-op and E2E serving benchmark data (Qwen3.8-27B-w8a8, single 910B3, CANN 9.0.0) plus accuracy runs are in the PR description above.

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_apply_top_k_top_p.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    torch.npu.synchronize()
+
+    assert logits_to_return is None
+    assert torch.equal(next_tokens, logits.argmax(dim=-1))
+
+
+def test_sampler_apply_path_processed_logits_matches_stock(sampler_module):
+    """processed_logits 模式返回 masked logits, 必须与 stock(CANN) 路径
+    位精确一致(逐请求混合 k/p)。"""
+    init_device_properties_triton()
```
Comment (weijinqian0) [inline tests/e2e/nightly/single_node/ops/singlecard_ops/test_apply_top_k_top_p.py:151]: plz use english.

File: vllm_ascend/envs.py [hunk 3]
```diff
@@ -87,6 +87,12 @@ def _strict_binary_env(name: str, default: str = "0") -> bool:
     # (safe for Ascend 910B/A3). Set to a positive value to override when
     # auto-detection is unavailable or for debugging UB overflow issues.
     "VLLM_ASCEND_ROPE_UB_SIZE_KB": lambda: int(os.getenv("VLLM_ASCEND_ROPE_UB_SIZE_KB") or 0),
+    # Whether to use the Triton apply_top_k_top_p op in sampling
+    # (vllm_ascend.ops.triton.apply_top_k_top_p: CANN npu_top_k_top_p
+    # semantics + optional fused softmax; takes per-request [B] k/p tensors
+    # directly, no .item() CPU sync). Performance-critical decode path.
+    # Default: 0 (disabled). Valid values: 0 or 1. Not sensitive.
+    "VLLM_ASCEND_USE_TRITON_APPLY_TOPK_TOPP": lambda: _strict_binary_env("VLLM_ASCEND_USE_TRITON_APPLY_TOPK_TOPP"),
 }
```
Comment (weijinqian0) [inline vllm_ascend/envs.py:96]: plz add this to addition_config.

File: vllm_ascend/ops/triton/apply_top_k_top_p.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+
+__all__ = [
+    "apply_top_k_top_p",
+    "apply_top_k_top_p_with_sorted",
+    "fused_topk_topp_softmax",
+    "fused_topk_topp_softmax_with_sorted",
+]
+
+# Single-pass path upper bound: fp32 intermediates at BLOCK=8192 exceed the
+# 192KB UB.
```
Comment (weijinqian0) [inline vllm_ascend/ops/triton/apply_top_k_top_p.py:68]: Needs to be compatible with A5.
